Update readme (#154)
tc39/proposal-type-annotations: README.md

> > ⚠️ *This document has not been updated regurlarly. See TC39 meeting notes from 2022 ([March 29](https://github.com/tc39/notes/blob/main/meetings/2022-03/mar-29.md#types-as-comments-for-stage-1), [March 31](https://github.com/tc39/notes/blob/main/meetings/2022-03/mar-31.md#types-as-comments-continuation)) and 2023 ([March 22](https://github.com/tc39/notes/blob/main/meetings/2023-03/mar-22.md#type-annotations-proposal-update)) for the most up-to-date information.*

Reviewer: fyi you missed a typo here

```diff
- regurlarly
+ regularly
```

Author: ty :) fixed

![missing-features-js](https://user-images.githubusercontent.com/6939381/143012138-96b93204-c456-4ab5-bb63-2648187ab8a7.png)

> Additionally, TypeScript was currently listed as the 4th most-used language in [GitHub's *State of the Octoverse*](https://octoverse.github.com/), and on [Stack Overflow's Annual Developer Survey](https://insights.stackoverflow.com/survey/) it has been listed in both the top 4 most-loved languages since 2017 and the 10 most-used languages.

Reviewer 2: This is important enough to merit _not_ removing, I believe.

But over time, we anticipate there will be less need for developers to downlevel-compile.
Evergreen browsers have become more of the norm, and on the back-end, Node.js and Deno use very recent versions of V8.

> Over time, for many Static Type users, the only necessary step between writing code and running it will be to erase away type annotations.

Reviewer 3: ```suggestion
Over time, for many static type system users, the only necessary step between writing code and running it will be to erase away type annotations.
```

This comment convention is often found in build scripts, small web apps, server-side apps, and elsewhere where
the cost/benefit tradeoff of adding a build-tool is too high.

> Even when no type-checking diagnostics in place, the comment convention is still leveraged in editors supporting types in their underlying JavaScript editing experience.

Reviewer 4: ```suggestion
Even when no type-checking diagnostics put in place, the comment convention is still leveraged in editors supporting types in their underlying JavaScript editing experience.
```

This comment convention is often found in build scripts, small web apps, server-side apps, and elsewhere where
the cost/benefit tradeoff of adding a build-tool is too high.

> Even when no type-checking diagnostics put in place, the comment convention is still leveraged in editors supporting types in their underlying JavaScript editing experience.

Reviewer 5: Even when no type-checking diagnostics _are_ put in place
or
Even _with_ no type-checking diagnostics put in place

Reviewer 5: Nevermind, reading this again later with fresh eyes, the original wording seems fine.

}

> ```

Reviewer 6: I would also add an example of Flow's comment types: https://flow.org/en/docs/types/comments/

```
function f(value /*: MyAlias */) /*: boolean */ {
  return value.bar;
}
```

Author: > ## In readme:
> > Type Assertions
> 
> `as` is TypeScript specific. You can add an example from Flow for our casting syntax: https://flow.org/en/docs/types/casting/
> 
> ```
> (x: T);
> ```
> 
> > One can explicitly specify the type arguments of a generic function invocation or generic class instantiation [in TypeScript](https://www.typescriptlang.org/docs/handbook/2/functions.html#specifying-type-arguments).
> 
> You can do the same in Flow, you can make this language more generic. https://flow.org/en/docs/types/generics/
> 

The first two comments have already been addressed. I suggest opening an issue or pull request to address the remaining comments. Do you want me to help or open those?

On top of this, JSDoc comments only provide a subset of the feature set supported in TypeScript,
in part because it's difficult to provide expressive syntax within JSDoc comments.

Nevertheless, the JSDoc-based syntax remains useful, and the need for some form of type annotations in JavaScript was significant enough for the TypeScript team to invest in it and the community to create tooling to support type-checking using JSDoc<sup>[[1](https://www.npmjs.com/package/flow-jsdoc)], [[2](https://www.npmjs.com/package/babel-plugin-jsdoc-runtime-typecheck)]</sup>.

For these reasons, this proposal explores and expects to a larger syntax to appear as-is in JavaScript source files, interpreted as comments.

## Proposal

> > The following is a Stage 1 proposal. Please treat it as such, updates are expected and [feedback](https://github.com/tc39/proposal-type-annotations/issues/new) it's welcome. [TC39 stage process document](https://tc39.es/process-document/)

Reviewer 7: typo : feedback **is** welcome

```

In the example above, `x` is annotated with the type `string`.

> Tools doing static type analyses can utilize that type, and might choose to error on the statement `x = 100`;

Reviewer 4: ```suggestion
Tools doing static type analysis can utilize that type, and might choose to error on the statement `x = 100`;
```

```

In the example above, `x` is annotated with the type `string`.

> Tools doing static type analysis can utilize that type, and might choose to error on the statement `x = 100`;

Reviewer 8: Static analysis tools can utilize that type....

(active over passive voice)

Reviewer 9: (I don't think that's active vs passive voice, an adverbial phrase that could be rewritten "tools, such as typescript, can ....")

### Kinds of Types

> The above examples use type names like `string`, `number`, and `boolean`, Most type-checkers support types with more involved syntax than just a single identifier.

Reviewer 4: ```suggestion
The above examples use type names like `string`, `number`, and `boolean`, the majority of static type-checkers support types with more involved syntax than just a single identifier.
```

> #### Non-Nullable Assertions

Reviewer 6: This whole feature is TypeScript specific. We should remove this whole section.

Author: As the currently suggested tentative grammar supports it and there is still room for discussion, I suggest we leave it as it is for now. However, we can address this topic in issues https://github.com/tc39/proposal-type-annotations/issues/31 and #142.
